cocos/core/gfx/webgl/webgl-device.ts

```diff
@@ -86,16 +87,33 @@ export class WebGLDevice extends Device {
         return this._swapchain!.nullTexCube;
     }
 
+    get bindingMappings () {
+        return this._bindingMappings!;
+    }
+
     private _swapchain: WebGLSwapchain | null = null;
     private _context: WebGLRenderingContext | null = null;
+    private _bindingMappings: IWebGLBindingMapping | null = null;
 
     public initialize (info: DeviceInfo): boolean {
         WebGLDeviceManager.setInstance(this);
         this._gfxAPI = API.WEBGL;
 
-        this._bindingMappingInfo = info.bindingMappingInfo;
-        if (!this._bindingMappingInfo.bufferOffsets.length) this._bindingMappingInfo.bufferOffsets.push(0);
-        if (!this._bindingMappingInfo.samplerOffsets.length) this._bindingMappingInfo.samplerOffsets.push(0);
+        const mapping = this._bindingMappingInfo = info.bindingMappingInfo;
+        const blockOffsets: number[] = [];
+        const samplerTextureOffsets: number[] = [];
+        for (let i = 0; i < mapping.setIndices.length; ++i) {
+            const cur = mapping.setIndices[i];
+            const last = i ? mapping.setIndices[i - 1] : cur;
```

Comment:
Don't understand the last here, should be `notFirst` ?
The logic in this for loop seems very loose, please add more comment to explain

Reply:
'last' as in the previous one, I'll add more comments about this